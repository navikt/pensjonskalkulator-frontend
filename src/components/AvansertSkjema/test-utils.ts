export function getPreviousMonth(): string {
  const today = new Date()
  const currentMonth = parseInt(
    today.toLocaleDateString('en-CA').slice(5, 7),
    10
  ) // Extracts the month from the date string
  const previousMonth = currentMonth - 1
  return previousMonth.toString()
}
