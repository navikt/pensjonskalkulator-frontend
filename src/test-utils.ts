import renderer from 'react-test-renderer'

import { render } from '@testing-library/react'

export function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
}

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => children,
    ...options,
  })

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
