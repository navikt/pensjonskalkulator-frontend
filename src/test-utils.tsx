import React, { PropsWithChildren } from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

import { PreloadedState, createListenerMiddleware } from '@reduxjs/toolkit'
import { render, RenderOptions } from '@testing-library/react'

import { getTranslation_test } from '@/utils/__tests__/test-translations'

import { createUttaksalderListener } from './state/listeners/uttaksalderListener'
import {
  setupStore,
  RootState,
  AppStore,
  AppStartListening,
} from './state/store'
import { getTranslation_nb } from './translations/nb'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
  hasRouter?: boolean
}

export const swallowErrors = (testFn: () => void) => {
  const error = console.error
  console.error = () => {}
  testFn()
  console.error = error
}

export const swallowErrorsAsync = async (testFn: () => Promise<void>) => {
  const cache = console.error
  console.error = () => {}
  await testFn()
  console.error = cache
}

function generateMockedTranslations() {
  const nbTranslations: Record<string, string> = getTranslation_nb()
  const testTranslations: Record<string, string> = getTranslation_test()
  const translationsInput = { ...nbTranslations, ...testTranslations }
  const translations: Record<string, string> = {}

  for (const key in translationsInput) {
    if (
      /<(?=.*? .*?\/ ?>|br|hr|input|!--|wbr)[a-z]+.*?>|<([a-z]+).*?<\/\1>|{[^}]+}/i.test(
        translationsInput[key]
      )
    ) {
      // for html or chunks keys: results in 'my_key' : 'my_key with some <html> or {chunk}'
      translations[key] = translationsInput[key]
    } else {
      // results in 'my_key' : 'my_key'
      translations[key] = key
    }
  }
  return translations
}

// Return an object with the store and all of RTL's query functions
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState, true),
    hasRouter = true,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <IntlProvider locale="nb" messages={generateMockedTranslations()}>
          {hasRouter ? <MemoryRouter>{children}</MemoryRouter> : children}
        </IntlProvider>
      </Provider>
    )
  }

  const listenerMiddleware = createListenerMiddleware()
  createUttaksalderListener(
    listenerMiddleware.startListening as AppStartListening
  )

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'

export { renderWithProviders as render }
