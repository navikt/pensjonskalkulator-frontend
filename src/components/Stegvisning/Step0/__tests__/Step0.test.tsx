import * as ReactRouterUtils from 'react-router'

import { describe, it, vi } from 'vitest'

import { Step0 } from '..'
import { render, screen, waitFor, fireEvent } from '@/test-utils'

describe('Step', () => {
  it('rendrer slik den skal, med riktig heading, bilde, tekst og knapper', async () => {
    const result = render(<Step0 />)

    await waitFor(() => {
      expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent(
        'stegvisning.steg0.title'
      )
      expect(result.asFragment()).toMatchSnapshot()
    })
  })

  it('sender videre til steg 1 n책r brukeren klikker p책 Neste', () => {
    const navigateMock = vi.fn()
    vi.spyOn(ReactRouterUtils, 'useNavigate').mockImplementation(
      () => navigateMock
    )
    render(<Step0 />)
    fireEvent.click(screen.getByText('stegvisning.steg0.neste_button'))
    expect(navigateMock).toHaveBeenCalledWith('/stegvisning/1')
  })

  it('redirigerer til landingssiden n책r brukeren klikker p책 Avbryt', () => {
    const navigateMock = vi.fn()
    vi.spyOn(ReactRouterUtils, 'useNavigate').mockImplementation(
      () => navigateMock
    )
    render(<Step0 />)
    fireEvent.click(screen.getByText('stegvisning.steg0.avbryt_button'))
    expect(navigateMock).toHaveBeenCalledWith('/')
  })
})
