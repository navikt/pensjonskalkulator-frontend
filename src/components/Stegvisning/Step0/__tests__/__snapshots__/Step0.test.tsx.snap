// Vitest Snapshot v1

exports[`Step > rendrer slik den skal, med riktig heading, bilde, tekst og knapper 1`] = `
<DocumentFragment>
  <section
    class="_section_b0eca1"
  >
    <img
      alt="stegvisning.steg0.bildetekst"
      class="_image_b0eca1"
      src="/pensjon/kalkulator/src/assets/frida.svg"
    />
    <h2
      class="navds-heading navds-heading--large navds-typo--spacing"
    >
      stegvisning.steg0.title
    </h2>
    <p
      class="_ingress_b0eca1 navds-body-short"
    >
      stegvisning.steg0.ingress
    </p>
    <button
      class="_button_b0eca1 navds-button navds-button--primary navds-button--medium"
    >
      <span
        aria-live="polite"
        class="navds-label"
      >
        stegvisning.steg0.neste_button
      </span>
    </button>
    <button
      class="navds-button navds-button--tertiary navds-button--medium"
    >
      <span
        aria-live="polite"
        class="navds-label"
      >
        stegvisning.steg0.avbryt_button
      </span>
    </button>
  </section>
</DocumentFragment>
`;
